File: JavaScript/Day13.js
(function(){
	"use strict";
	
	require("fs").readFile("../inputs/Day13.txt", "utf8", function(err, data){
		if(err){
			console.log(`Unable to load file: ${err}`);
		} else {
			let people = [], happiness = {};

			data.split('\n').map(l => {
				let parts = l.match(/([a-z]+) would (gain|lose) ([0-9]+) happiness units by sitting next to ([a-z]+)\./i), person = parts[1], amount = (parts[2] == 'gain' ? 1 : -1) * parts[3], neighbor = parts[4];
				if(people.indexOf(person) === -1){
					people.push(person);
				}
				if(typeof happiness[person] === 'undefined'){
					happiness[person] = {};
				}
				happiness[person][neighbor] = amount;
			});

			let originalBest = findBestArrangement(people, happiness);
			
			happiness['self'] = {};
			for(let person of people){
				happiness[person]['self'] = 0;
				happiness['self'][person] = 0;
			}
			people.push('self');

			let includingMeBest = findBestArrangement(people, happiness);

			console.log(`Part 1: ${originalBest}`);
			console.log(`Part 2: ${includingMeBest}`);
		}
	});

	let findBestArrangement = function(people, happiness) {
		let best = 0;
		let arrangements = permute(people);
		for(let seating of arrangements){
			let total = 0, seatCount = seating.length;
			for(let s = 0; s < seatCount; s++){
				let left = s - 1;
				if(left < 0){
					left = seatCount - 1;
				}

				let right = s + 1;
				if(right >= seatCount){
					right = 0;
				}

				total += happiness[ seating[s] ][ seating[left] ] + happiness[ seating[s] ][ seating[right] ];
			}
			best = Math.max(total, best);
		}
		return best;
	}

	let permute = function(list) {
		if(list.length === 0){
			return [[]];
		}

		let permutations = [];
		for(let l = 0; l < list.length; l++){
			let copy = Object.create(list);
			let head = copy.splice(l, 1);
			let remaining = permute(copy);
			for(let tail of remaining){
				permutations.push(head.concat(tail));
			}
		}
		return permutations;
	};
})();

File: Python/Day13.py
from itertools import permutations
import re

rules = open('../inputs/Day13.txt').read().split('.\n')

def findBestArrangement(people, happiness):
	best = 0
	for seating in permutations(people):
		total = 0
		for s in xrange(0, len(seating)):
			left = s - 1
			if left < 0:
				left = len(seating) - 1
			right = s + 1
			if right >= len(seating):
				right = 0

			total += happiness[(seating[s], seating[left])] + happiness[(seating[s], seating[right])]

		best = max(best, total)
	return best

people = set()
happiness = {}

for rule in rules:
	parts = re.search('(\w+) would (\w+) (\d+) happiness units by sitting next to (\w+)', rule).groups()
	person = parts[0]
	neighbor = parts[3]
	amount = int(parts[2]) * (1 if parts[1] == 'gain' else -1)
	people.add(person)
	happiness[(person, neighbor)] = amount

originalBest = findBestArrangement(people, happiness)

for person in people:
	happiness[(person, 'self')] = 0
	happiness[('self', person)] = 0
people.add('self')

includingMeBest = findBestArrangement(people, happiness)

print "Part 1:", originalBest
print "Part 2:", includingMeBest